### ISSUE TITLE
padding-line-between-statements: Custom STATEMENT_TYPE

### ISSUE THREAD

**Issue by jadengore** [eslint/eslint#8849]
**What rule do you want to change?**
`padding-line-between-statements`

**Does this change cause the rule to produce more or fewer warnings?**
Fewer warnings.

**How will the change be implemented? (New option, new default behavior, etc.)?**
Allow specification of a custom `STATEMENT_TYPE` that allows a user to specify certain statements to require spacing, such as an `it()` statement in Mocha/Jest.

**Please provide some example code that this change will affect:**
Before allowing custom `STATEMENT_TYPE`:
```js
describe("Test Suite", () => {
    it("has a test", () => {
        expect(1).to.equal(1);
    });
    it("has another test", () => { // No way to enforce an error here
        expect(2).to.equal(2);
    });
});
```
After allowing a custom `STATEMENT_TYPE` that adds `it` as a statement requiring one or more newlines:
```js
describe("Test Suite", () => {
    it("has a test", () => {
        expect(1).to.equal(1);
    });
    it("has another test", () => { // Error
        expect(2).to.equal(2);
    });
});
```
The code would be valid with the added option if it looked like this:
```js
describe("Test Suite", () => {
    it("has a test", () => {
        expect(1).to.equal(1);
    });

    it("has another test", () => { // Valid
        expect(2).to.equal(2);
    });
});
```

**What does the rule currently do for this code?**
Nothing.

**What will the rule do after it's changed?**
Enforce padding line between a custom `STATEMENT_TYPE`.

**Comment by not-an-aardvark** [comment]
Thanks for the proposal. What do you mean by a "custom `STATEMENT_TYPE`"? Could you give an example of what the new options would look like?

**Comment by jadengore** [comment]
So, currently a `STATEMENT_TYPE` to me looks like a bunch of Javascript language keywords/tokens. What I mean by `custom` is allow a custom token type for the statement, such as `it`. 

My initial idea for it looked something like this:
```json
{
    "padding-line-between-statements": [
        "error",
        { "blankLine": "always", "prev": "it", "next": "it" }
    ]
}
```
However, I'm realizing this is actually an expression type rather than a statement. Maybe requiring padding after certain function calls?

**Comment by platinumazure** [comment]
Can we support selectors here?

For example: `{ blankLine: "always", prev: "ExpressionStatement > CallExpression[callee.name='it']", next: "ExpressionStatement > CallExpression[callee.name='it']" }`

...A bit verbose, though!

**Comment by jadengore** [comment]
@platinumazure That's a great idea! Didn't think of that. The key is that the code here for STATEMENT_TYPE only recognizes keyword tokens from Esprima, but if we could specify tokens that represent the expressions this could work.

Would we have to limit to `ExpressionStatement` and `CallExpression`? Wonder what the other use cases could be here.

**Comment by platinumazure** [comment]
Well, I was thinking based on the rule name, we care mainly about padded lines between statements rather than expressions or parts of an expression. So I had assumed CallExpression nodes would only be relevant to this rule if they were immediate children of ExpressionStatement nodes. But maybe that understanding could be implicit and so the selector wouldn't need to start with a statement type.

I think the existing keyword-based statement types could easily be transformed into selectors behind the scenes, if we wanted.

**Comment by not-an-aardvark** [comment]
I don't think the keyword-based statement types can be transformed into selectors because some of them take formatting into account (e.g. "multiline-block-type" checks whether something is split on multiple lines).

**Comment by jadengore** [comment]
@not-an-aardvark Maybe we are considering this the wrong way... It might just have to be an entirely separate rule to support this. `padding-line-between-expressions`?

**Comment by not-an-aardvark** [comment]
Hmm, I think that's unnecessary -- it would conflict with the existing `padding-line-between-statements` rule.

It would be possible to add selector options to the `padding-line-between-statements` rule, which might satisfy this use case -- there would just be a caveat that they wouldn't be able to express everything that the existing string options can express.

**Comment by jadengore** [comment]
@not-an-aardvark Can you give me an example?

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.
